### ISSUE TITLE
Mask RCNN: get scores for all 90 classes of a single box

### ISSUE THREAD

**Issue by dongmingsun** [tensorflow/models#4637]
### Describe the problem
I know this is **NOT** a bug but few people on stackoverflow know the solution. So I didn't fill out the information required.

I'm using the `mask_rcnn_inception_v2_coco_2018_01_28` model for instance segmentation, and the model outputs detection_scores with dims=[1, 100], which corresponds to the 100 boxes.

While I wonder if I can obtain scores with dims=[90, 100], i.e., for each box I need 90 scores for all the classes, rather than just the highest score.

Similarly, I also want the masks with dims=[90, 100, 15, 15], rather than [1, 100, 15, 15].

I tried to find the solution by examining the model (I visualise the model using tensorboard, the files is on [GoogleDrive](https://drive.google.com/file/d/1hxoo7ItwinLUNf79EvLBFae-rGYpqnTp/view?usp=sharing)). But is really confused by the graph.

### Source code / logs
The code is [here](https://github.com/tensorflow/models/blob/master/research/object_detection/object_detection_tutorial.ipynb). And I'm using the C++ API to do the inference:

```c++
Status run_status = session->Run({{input_layer, resized_tensor}},
                                   {"num_detections:0", "detection_boxes:0",
                                    "detection_scores:0", "detection_classes:0",
                                    "detection_masks:0"},
                                   {}, &outputs); // original:{output_layer}
```

with output:

```
==============================
detection_scores:0
Tensor<type: float shape: [1,100] values: [0.998502731 0.986097693 0.821735203]...>

==============================
detection_classes:0
Tensor<type: float shape: [1,100] values: [18 18 1]...>

==============================
detection_masks:0
Tensor<type: float shape: [1,100,15,15] values: [[[0.000649190042 0.000243070099 0.000696995761]]]...>
```

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
What is the top-level directory of the model you are using
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by pierrejeambrun** [comment]
Hi @dongmingsun,

I also needed all the scores to do further post processing. I am using the faster_rcnn model for object detection so I don't know if this can be applied to your specific use case.

I believe that the function responsible for the scores is the `"score_converter_fn"` that is passed to the **MetaArch** by the `model_builder`. You should see it in your **config** file. In my case it is in the `"second_stage_postprocessing/score_converter"` section (in most cases SOFTMAX but default is IDENTITY in the proto file). Then retrieve the corresponding tensor.

Something like:
``` python
all_scores = detection_graph.get_tensor_by_name("SecondStagePostprocessor/Softmax")
```
This gives a tensor of shape [nb_detection, nb_classes + 1]. (0 being the background)

I don't think there is a standardized endpoint yet and it may vary depending on your config file.

Hope this helps.

**Comment by dongmingsun** [comment]
Hi @pierrejeambrun,

Thanks a lot, your methods works. Thankfully the authors of the model made the tensor name meaningful.

Actually the scores and masks I need are tensors named `"Softmax_1:0"` and `"SecondStageBoxPredictor_1/MaskPredictor:0"` respectively.

**Comment by Jacquelinemanlho** [comment]
Hi @dongmingsun I would like to ask would you mind to share how to extract the mask as I got issue on it. 

more information: when I run in .py it return the shape of detection_masks is (94, 324, 576); however in C++ tensorflow API it return detection_masks: Tensor<type: float shape: [1,300,15,15]

I had tried "SecondStageBoxPredictor_1/MaskPredictor:0" and verified the data compare with detection_masks. It is same.

===================================
example: (SecondStageBoxPredictor_1/MaskPredictor:0)
output[0](0, 0, 0, 0, 0) = 0.369476
output[0](0, 0, 0, 0, 1) = 0.258394
output[0](0, 0, 0, 0, 2) = 0.878370
output[0](0, 0, 0, 0, 3) = 0.779922
output[0](0, 0, 0, 0, 4) = 0.412113
output[0](0, 0, 0, 0, 5) = 0.277506
output[0](0, 0, 0, 0, 6) = 0.910097
output[0](0, 0, 0, 0, 7) = 0.619292
output[0](0, 0, 0, 0, 8) = 0.128057
output[0](0, 0, 0, 0, 9) = 0.277343
output[0](0, 0, 0, 0, 10) = 0.801713
output[0](0, 0, 0, 0, 11) = 0.826745
output[0](0, 0, 0, 0, 12) = 0.417097
output[0](0, 0, 0, 0, 13) = 0.361232
output[0](0, 0, 0, 0, 14) = 0.923316

output[0](0, 0, 0, 1, 0) = 0.258394
output[0](0, 0, 0, 1, 1) = 0.878370
output[0](0, 0, 0, 1, 2) = 0.779922
output[0](0, 0, 0, 1, 3) = 0.412113
output[0](0, 0, 0, 1, 4) = 0.277506
output[0](0, 0, 0, 1, 5) = 0.910097
output[0](0, 0, 0, 1, 6) = 0.619292
output[0](0, 0, 0, 1, 7) = 0.128057
output[0](0, 0, 0, 1, 8) = 0.277343
output[0](0, 0, 0, 1, 9) = 0.801713
output[0](0, 0, 0, 1, 10) = 0.826745
output[0](0, 0, 0, 1, 11) = 0.417097
output[0](0, 0, 0, 1, 12) = 0.361232
output[0](0, 0, 0, 1, 13) = 0.923316
output[0](0, 0, 0, 1, 14) = 0.723455
===================================

===================================
detection_masks
output[0](0, 0, 0, 0) = 0.369476
output[0](0, 0, 0, 1) = 0.258394
output[0](0, 0, 0, 2) = 0.878370
output[0](0, 0, 0, 3) = 0.779922
output[0](0, 0, 0, 4) = 0.412113
output[0](0, 0, 0, 5) = 0.277506
output[0](0, 0, 0, 6) = 0.910097
output[0](0, 0, 0, 7) = 0.619292
output[0](0, 0, 0, 8) = 0.128057
output[0](0, 0, 0, 9) = 0.277343
output[0](0, 0, 0, 10) = 0.801713
output[0](0, 0, 0, 11) = 0.826745
output[0](0, 0, 0, 12) = 0.417097
output[0](0, 0, 0, 13) = 0.361232
output[0](0, 0, 0, 14) = 0.923316

output[0](0, 0, 1, 0) = 0.258394
output[0](0, 0, 1, 1) = 0.878370
output[0](0, 0, 1, 2) = 0.779922
output[0](0, 0, 1, 3) = 0.412113
output[0](0, 0, 1, 4) = 0.277506
output[0](0, 0, 1, 5) = 0.910097
output[0](0, 0, 1, 6) = 0.619292
output[0](0, 0, 1, 7) = 0.128057
output[0](0, 0, 1, 8) = 0.277343
output[0](0, 0, 1, 9) = 0.801713
output[0](0, 0, 1, 10) = 0.826745
output[0](0, 0, 1, 11) = 0.417097
output[0](0, 0, 1, 12) = 0.361232
output[0](0, 0, 1, 13) = 0.923316
output[0](0, 0, 1, 14) = 0.723455

**Comment by dongmingsun** [comment]
@Jacquelinemanlho Hi, I think you have two questions right? 

**First**, if you want to retrieve masks, you would want to find the name of the tensor, which is `"detection_masks:0"`. I think you've already done this part, the snippet below is just to make sure we have the same thing here :

```cpp
tensorflow::Status tf_run_status =
                tf_session->Run({{tf_input_layer, tf_image_tensor}},
                                {
                                        "num_detections:0",    // outputs[0]
                                        "detection_boxes:0",   // outputs[1]
                                        "detection_scores:0",  // outputs[2]
                                        "detection_classes:0", // outputs[3]
                                        "detection_masks:0"    // outputs[4]},
                                },
                                {}, &tf_outputs);
```

**Second**, you mentioned the size of the mask. I believe the size in the python code is something after post-processed, so if you examine the raw output it would be the same as in the C++ code.

Well, I can't recall if I had checked the size, but I think the size you [1,300,15,15] is right and it's just a fixed-sized mask which should be resize and re-projected back to the input image, below is the C++ function I wrote before, and I hope it's useful to you.

```cpp
tensorflow::Status ReframeBoxMasksToImageMasks(tensorflow::Tensor &intput_masks,
                                   tensorflow::Tensor &input_boxes,
                                   int num_detections,
                                   std::vector<tensorflow::Tensor> *out_tensors, int height,
                                   int width) {
  /*
   *intput_masks: [1,100,15,15]
   *input_boxes: [1,100,4]
   */
  tensorflow::string output_name = "output";
  auto root = tensorflow::Scope::NewRootScope();
  using namespace ::tensorflow::ops;

  // Prepare masks for CropAndResize
  // Shape: [6,15,15]
  auto masks_reduced =
      Slice(root, intput_masks, {0, 0, 0}, {num_detections, 15, 15});
  // Shape: [6,15,15,1]
  auto masks_reshaped = ExpandDims(root, masks_reduced, -1);
//      Reshape(root, masks_reduced, {num_detections, 15, 15, 1})
  // Prepare masks for CropAndResize done

  // Prepare boxes for CropAndResize
  // Shape: [1,6,2]
  auto input_boxes_expanded = ExpandDims(root, input_boxes, 0);
  auto min_corner = Slice(root, input_boxes_expanded, tensorflow::Input{0, 0, 0},
                          tensorflow::Input{1, num_detections, 2});
  auto max_corner = Slice(root, input_boxes_expanded, tensorflow::Input{0, 0, 2},
                          tensorflow::Input{1, num_detections, 2});
  // Shape: [6,1,2]
  auto min_corner_reshaped = Reshape(root, min_corner, {num_detections, 1, 2});
  auto max_corner_reshaped = Reshape(root, max_corner, {num_detections, 1, 2});

  // Prepare unit boxes
  tensorflow::TensorShape unit_boxes_shape({num_detections, 2});
  tensorflow::Tensor unit_boxes_zeros(tensorflow::DT_FLOAT, unit_boxes_shape);
  tensorflow::Tensor unit_boxes_ones(tensorflow::DT_FLOAT, unit_boxes_shape);
  for (int row = 0; row < num_detections; ++row) {
    for (int col = 0; col < 2; ++col) {
      unit_boxes_zeros.tensor<float, 2>()(row, col) = 0;
    }
  }
  for (int row = 0; row < num_detections; ++row) {
    for (int col = 0; col < 2; ++col) {
      unit_boxes_ones.tensor<float, 2>()(row, col) = 1;
    }
  }
  auto unit_boxes = Concat(root, {unit_boxes_zeros, unit_boxes_ones}, 1);
  // Unit boxes shape: [6,2,2]
  auto unit_boxes_reshaped = Reshape(root, unit_boxes, {-1, 2, 2});
  auto transformed_boxes =
      Div(root, Sub(root, unit_boxes_reshaped, min_corner_reshaped),
          Sub(root, max_corner_reshaped, min_corner_reshaped));
  auto transformed_boxes_reshaped = Reshape(root, transformed_boxes, {-1, 4});
  // Prepare boxes for CropAndResize done

  // Prepare box_ind for CropAndResize
  tensorflow::TensorShape box_ind_shape({num_detections});
  tensorflow::Tensor box_ind(tensorflow::DT_INT32, box_ind_shape);
  auto box_ind_vec = box_ind.vec<tensorflow::int32>();
  for (int i = 0; i < num_detections; ++i) {
    box_ind_vec(i) = i;
  }

  // Produce the masks reframed to the image we need
  auto masks_reframed =
      CropAndResize(root, masks_reshaped, transformed_boxes_reshaped, box_ind,
                    {height, width});
//  auto masks_squeezed = Squeeze(root.WithOpName(output_name), masks_reframed);
  auto masks_squeezed = Reshape(root.WithOpName(output_name), masks_reframed, {-1, 480, 640});

  // This runs the GraphDef network definition that we've just
  // constructed, and returns the results in the output tensor.
  tensorflow::GraphDef graph;
  TF_RETURN_IF_ERROR(root.ToGraphDef(&graph));
  std::unique_ptr<tensorflow::Session> session(
      tensorflow::NewSession(tensorflow::SessionOptions()));
  TF_RETURN_IF_ERROR(session->Create(graph));
  TF_RETURN_IF_ERROR(session->Run({}, {"output"}, {}, out_tensors));
  return tensorflow::Status::OK();
}

tensorflow::Status SaveAsJPG(const tensorflow::Tensor &input, tensorflow::string file_name) {
  /*
   *input: [1,height,width,3]
   */
  auto root = tensorflow::Scope::NewRootScope();
  using namespace ::tensorflow::ops;
  // Shape: [height, width, 3]
  // @input_squeezed shape: [height, width, 3]
  auto input_squeezed = Squeeze(root, input);
  // @input_encoded: 0-D. JPEG-encoded image.
  auto input_encoded = EncodeJpeg(root, input_squeezed);
  auto created_operation =
      WriteFile(root.WithOpName("output/image"), file_name, input_encoded);

  // Run the GraphDef network
  // std::vector<tensorflow::Tensor> *out_tensors;
  tensorflow::GraphDef graph;
  TF_RETURN_IF_ERROR(root.ToGraphDef(&graph));
  std::unique_ptr<tensorflow::Session> session(
      tensorflow::NewSession(tensorflow::SessionOptions()));
  TF_RETURN_IF_ERROR(session->Create(graph));
  TF_RETURN_IF_ERROR(session->Run({}, {}, {"output/image"}, {}));
  return tensorflow::Status::OK();
}
```
